learn-typescript-linting.md

```diff
@@ -0,0 +1,503 @@
+
+```bash
+npm i -D eslint @typescript-eslint/{eslint-plugin,parser}
+touch .eslintrc.json
+```
+
+### 2.1 Initializing .eslintrc
+[`[toc]`](#table-of-contents)
+
+Add the following to `.eslintrc.json`
+
+```json
+{
+  "extends": ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
+  "parser": "@typescript-eslint/parser",
+  "plugins": ["@typescript-eslint"],
+  "env": { "node": true },
+  "parserOptions": {
+    "ecmaVersion": 5,
+    "sourceType": "module"
+  },
+  "rules": {}
+}
+```
+
+#### 2.1.1 Specifying environments
+[`[toc]`](#table-of-contents)
+
+The `env` object is used for defining global variables in a project that are not explicitly imported. A great example is the `console` object. This is considered globally available in both browser and Node.js environments. In the `learn-typescript-linting` example, the code will be executed in a Node.js terminal, thus the `node` attributes is enabled. Some other common attributes include `jest`, `mocha`, `amd`, `commonjs`, and `es6`. There is no easy way to know which attributes need to be enabled; it is recommended to consult [ESLint's Specifying Environments](https://eslint.org/docs/user-guide/configuring#specifying-environments) documentation to find out what each environment attribute provides.
+
+#### 2.1.2 Specifying ecmaVersion
+[`[toc]`](#table-of-contents)
+
+The `parserOptions.ecmaVersion` value is based on the `target` value found in the `tsconfig.json`. A `tsconfig.json` value of `{ "target": "es5" }` is equivalent to `{ "ecmaVersion": 5 }`. Use the table below for additional mappings.
+
+| tsconfig.json `target` | .eslintrc.json `ecmaVersion` |
+| ------ | ----------- |
+| `es3` | 3 |
+| `es5` | 5 |
+| `es6` or `ES2015` | 6 or 2015 |
+| `ES2016` | 7 or 2016 |
+| `ES2017` | 8 or 2017 |
+| `ES2018` | 9 or 2018 |
+| `ES2019` | 10 or 2019 |
+| `ESNext` | 10 or 2019 |
+
+Take a look at TypeScript's `--lib` [compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html) to learn how to include unique library files in the compilation. By setting `target` to either `es5` or `es6`, TypeScript will automatically import a set of libraries (i.e. `{ target: es5 } = { lib: ['DOM', 'ES5', 'ScriptHost']}`).
+
+### 2.2 Creating ESLint npm script
+[`[toc]`](#table-of-contents)
+
+Now that ESLint is configured, create a new npm script in `package.json` by adding it to the `"scripts"` object:
+
+```diff
+{
+  "scripts": {
+    "compile": "tsc -p tsconfig.json",
+    "start": "node lib/index.js",
++   "lint": "eslint 'src/**/*.ts'"
+  }
+}
+```
+
+This command will run ESLint on all `.ts` files within the `src` directory. The `/**/*` glob pattern will reach all files within subdirectories of `src`. If you have multiple directories to run the linter on (such as a `test` directory), use a pattern such as: `{src,test}/**/*.ts`.
+
+### 2.3 Executing ESLint
```

Comment:
Although this is correct it does not work on windows for some reason and after some searching I do not know if there is a way to get it to work on windows, the way I made it work is by using a Windows Subsystem for Linux and there were a couple of things I had to do to get that to work which I can expand upon if needed.

Reply:
I got it to work by getting rid of the quotes:
```diff
- "lint": "eslint 'src/**/*.ts'"
+ "lint": "eslint src/**/*.ts"
```

Reply:
Without quotes seems to work on Mac too

Reply:
Yes

Reply:
Okay I pushed changes to all necessary branches for this change. I'll update the article accordingly too